build-pyspy.bat

```diff
@@ -0,0 +1,35 @@
+@echo off
+WHERE cargo
+IF %ERRORLEVEL% NEQ 0 (
+	ECHO cargo wasn't found. Attempting to install...
+	.\dep\rustup-init.exe -y
```

Comment:
Instead of comitting it to the repo - download it during build as decribed in https://forge.rust-lang.org/infra/other-installation-methods.html#other-ways-to-install-rustup. Eventually this bat file is meant to run in the CICD on Windows, so I assume there IS a way to download & install external dependencies (i.e either CIs come with e.g `wget` installed, or you have some way to specify URLs you wish to be downloaded.

I checked [micropython](https://github.com/micropython/micropython) which has a [Windows CI](https://github.com/micropython/micropython/blob/master/ports/windows/.appveyor.yml), we'll target something like that, you can check if AppVeyor images have wget or curl and then we can depend on that.

Same for other dependencies - refrain from committing them and prefer to get them during CI / this script

Reply:
Ok, makes sense. I thought the batch script had to be self-contained without expecting the user to install any dependency. I can strip out the binaries for Rust and Windows SDK/Visual C++ build tools and assume they are already installed in your CI environment. Are you fine with this?

Reply:
The script needn't be self container. It's fair to download dependencies here (and it makes no sense to commit them - best to download and by that the user sees exactly what gets packed inside)

Reply:
For Rust, you'll need to install a wget executable to then pull the Rust exe. For visual studio the executable that is included in the dep folder is what installs the needed components. At some point, we have to make assumptions about what dependencies are pre-installed in your CI environment - whether it is wget/rust/visual studio build tools or all of them. By the way, these dependencies are only needed to build the py-spy.exe. Can we assume py-spy.exe is built and available in your CI environment?

Reply:
We can assume that wget is installed (i.e require the user to install it). We can't assume py-spy is available - we need to build it, mostly because we build it from a specific revision which includes our changes (see `scripts/pyspy_build.sh`).

Let's prepare that list of required dependencies? If VS SDK & Rust can be installed automatically once you have wget - that's good, and let's require only wget.

Reply:
See dependencies recommendation below

> ### **Dependencies**
> > **Options**
> 
> 1. [Recommended]: Provide prebuilt executable for py-spy on Windows and build gprofiler executable with provided build script. You control the build environment for the py-spy and the correct tags
> 2. Assume that the following prerequisites are available on target machine:
> 
> * Python (>=3.9.13)
> * Pip (>=22.2.2)
> * Rust Compiler
> * py-spy source with appropriate tag and version
> * VS Build Tools executable
> 
> > Using VS Build Tools install following components
> 
> * Microsoft.VisualStudio.Workload.VCTools
> * Microsoft.VisualStudio.Component.Windows10SDK.19041
> * Microsoft.VisualStudio.Component.VC.Tools.x86.x64
> 
> > Build py-spy and gprofiler executables with provided build script
> 
> 3. Assume that the following prerequisites are available on target machine:
> 
> * Python (>=3.9.13)
> * Pip (>=22.2.2)
> * py-spy source with appropriate tag and version
> * Wget
> 
> > Using wget download and install:
> 
> * Rust compiler
> * VS Build Tools executable
> 
> > Using VS Build Tools executable, install the following components:
> 
> * Microsoft.VisualStudio.Workload.VCTools
> * Microsoft.VisualStudio.Component.Windows10SDK.19041
> * Microsoft.VisualStudio.Component.VC.Tools.x86.x64
> 
> > Build py-spy and gprofiler executables with provided build script
> 
> 4. Assume that the following prerequisites are available on target machine:
> 
> * Python (>=3.9.13)
> * Pip (>=22.2.2)
> * Git bash (https://community.chocolatey.org/packages/git)
> * Wget
> 
> > Using Git, clone py-spy source with appropriate tag and version
> > Using wget, download and install:
> 
> * Rust compiler
> * VS Build Tools executable
> 
> > Using VS Build Tools executable, install the following components:
> 
> * Microsoft.VisualStudio.Workload.VCTools
> * Microsoft.VisualStudio.Component.Windows10SDK.19041
> * Microsoft.VisualStudio.Component.VC.Tools.x86.x64
> 
> > Build py-spy and gprofiler executables with provided build script

Reply:
Closing the discussion here, we continue in comments